lib/dialect/src/ConstraintsAnalysis.cpp

```diff
@@ -0,0 +1,1229 @@
+		return to_check.contains(value) and this->ranges.empty();
+	}
+
+	// Useful case for branch checking
+	bool ConstraintsLattice::isOnlyToCheckPresentInLattice(const mlir::Value& value) const{
+		
+		for(auto pair : this->ranges){
+			if(pair.first.first==value)
+				return false;
+		}
+		return not this->to_check.empty();
+	}
+
+	// Useful case for branch checking
+	bool ConstraintsLattice::isValueOfTypePresentInLattice(bool type) const{
+
+		for(auto pair : this->ranges){
+			if(pair.first.second==type)
+				return true;
+		}
+		return false;
+	}
+
+	// Useful case for branch checking
+	bool ConstraintsLattice::rangesEmptyAndToCheckOneElement() const{
+		return this->ranges.empty() and this->to_check.size()==1;
+	}
+
+	// Useful method
+	bool ConstraintsLattice::rangesEmpty() const{
+		return this->ranges.empty();
+	}
+
+	// Useful method
+	bool ConstraintsLattice::toCheckEmpty() const{
+		return this->to_check.empty();
+	}
+
+	// Useful method
+	bool ConstraintsLattice::isInToCheck(const mlir::Value& value) const{
+		auto found=this->to_check.find(value);
+		return found!=this->to_check.end();
+	}
+
+	// Method called in arithmetic operations
+	bool ConstraintsLattice::findInterestingOperandAndUpdate(
+	mlir::Operation* op,
+	mlir::rlc::ConstraintsAnalysis* analysis, 
+	const mlir::Value& result){
+
+		// Keep operand aliases for clarity
+		const auto& op1=op->getOperand(0);
+		const auto& op2=op->getOperand(1);
+		const auto& op1_defop=op1.getDefiningOp();
+		const auto& op2_defop=op2.getDefiningOp();
+
+		// Understand the operand that I have to update
+		mlir::Value update=nullptr;
+		mlir::Value  other=nullptr;
+
+		// Check which operand should be updated (the other should be a constant)
+		// TODO: understand what to do when there are no constants
+		if(op1_defop == nullptr or not mlir::isa<mlir::rlc::Constant>(op1_defop)){
+			update=op1;
+			other=op2;
+		}
+		else if(op2_defop == nullptr or not mlir::isa<mlir::rlc::Constant>(op2_defop)){
+			update=op2;
+			other=op1;
+		}
+		else return false;
+
+		bool ret1=this->updateArithmeticRange(op, analysis, result, update, other, true);
+		bool ret2=this->updateArithmeticRange(op, analysis, result, update, other, false);
+
+		return ret1 or ret2;
+	}
+
+	// Method that checks the lattice and updates the value accordingly
+	bool ConstraintsLattice::updateArithmeticRange(
+		mlir::Operation* op,
+		mlir::rlc::ConstraintsAnalysis* analysis, 
+		const mlir::Value& result,
+		const mlir::Value& update,
+		const mlir::Value& other,
+		bool branch)
+	{
+
+		auto result_range = this->ranges.find(std::make_pair(result,branch));
+
+		// Corner casess
+		if(result_range==this->ranges.end() or result_range->second==this->getTOP()) return false;
+
+		ConstraintsLattice::IntegerRange new_range(ConstraintsLattice::APInteger(1),ConstraintsLattice::APInteger(0));
+		// If the other is a constant take its constant value
+		if(auto other_casted=mlir::dyn_cast<mlir::rlc::Constant>(other.getDefiningOp())){
```

Comment:
constraint lattice cannot know the exhaustive list of operations to handle here, because they may change.